instrumentation/rocketmq-client/src/main/java/brave/rocketmq/client/RocketMQTracing.java

```diff
@@ -0,0 +1,147 @@
+  final Extractor<TracingProducerRequest> producerExtractor;
+  final Extractor<TracingConsumerRequest> consumerExtractor;
+  final Injector<TracingProducerRequest> producerInjector;
+  final Injector<TracingConsumerRequest> consumerInjector;
+  final String[] traceIdHeaders;
+  final SamplerFunction<MessagingRequest> producerSampler, consumerSampler;
+  final String remoteServiceName;
+
+  RocketMQTracing(MessagingTracing messagingTracing,
+    String remoteServiceName) { // intentionally hidden constructor
+    this.tracing = messagingTracing.tracing();
+    this.tracer = tracing.tracer();
+    Propagation<String> propagation = messagingTracing.propagation();
+    this.producerExtractor = propagation.extractor(TracingProducerRequest.GETTER);
+    this.consumerExtractor = propagation.extractor(TracingConsumerRequest.GETTER);
+    this.producerInjector = propagation.injector(TracingProducerRequest.SETTER);
+    this.consumerInjector = propagation.injector(TracingConsumerRequest.SETTER);
+    this.producerSampler = messagingTracing.producerSampler();
+    this.consumerSampler = messagingTracing.consumerSampler();
+    this.remoteServiceName = remoteServiceName;
+
+    // We clear the trace ID headers, so that a stale consumer span is not preferred over current
+    // listener. We intentionally don't clear BaggagePropagation.allKeyNames as doing so will
+    // application fields "user_id" or "country_code"
+    this.traceIdHeaders = propagation.keys().toArray(new String[0]);
+  }
+
+  <R> TraceContextOrSamplingFlags extractAndClearTraceIdHeaders(Extractor<R> extractor,
+    R request,
+    Map<String, String> properties) {
+    TraceContextOrSamplingFlags extracted = extractor.extract(request);
+    // Clear any propagation keys present in the headers
+    if (extracted.samplingFlags() == null) { // then trace IDs were extracted
+      if (properties != null) {
+        clearTraceIdHeaders(properties);
+      }
+    }
+    return extracted;
+  }
+
+  /** Creates a potentially noop remote span representing this request. */
+  Span nextMessagingSpan(SamplerFunction<MessagingRequest> sampler, MessagingRequest request,
+    TraceContextOrSamplingFlags extracted) {
+    Boolean sampled = extracted.sampled();
+    // only recreate the context if the messaging sampler made a decision
+    if (sampled == null && (sampled = sampler.trySample(request)) != null) {
+      extracted = extracted.sampled(sampled);
+    }
+    return tracer.nextSpan(extracted);
+  }
+
+  // We can't just skip clearing headers we use because we might inject B3 single, yet have stale B3
+  // multi, or vice versa.
+  void clearTraceIdHeaders(Map<String, String> headers) {
+    for (String traceIDHeader : traceIdHeaders)
+      headers.remove(traceIDHeader);
+  }
+
+  public Tracing tracing() {
+    return tracing;
+  }
+
+  public Tracer tracer() {
+    return tracer;
+  }
+
+  public MessageListenerOrderly wrap(MessageListenerOrderly messageListenerOrderly) {
```

Comment:
for these functions, follow existing convention which is to use a function name that is the same as the type being wrapped. Also needs javadoc. You can look at kafka-tracing for example, but make sure when following the conventions, the javadoc says correct. e.g. if it will produce a span or if it is just for adding tags:

```java
  /**
   * Extracts or creates a {@link Span.Kind#CONSUMER} span for each message received. This span is
   * injected onto each message so it becomes the parent when a processor later calls {@link
   * #nextSpan(ConsumerRecord)}.
   */
  public <K, V> Consumer<K, V> consumer(Consumer<K, V> consumer) {
    if (consumer == null) throw new NullPointerException("consumer == null");
    return new TracingConsumer<>(consumer, this);
  }
```